File: Sorting Algorithms/C++/QuickSort.cpp
#include <iostream>

void printVector(std::vector<int> &vector){
    for(int i=0; i<vector.size(); i++){
        if(i==0){
            std::cout << "[";
        }
        if(i== vector.size()-1){
            std::cout << vector[i] << "]" << std::endl;
        } else {
            std::cout << vector[i] << ",";
        }
    }
}

void swap(std::vector<int> &vector, int i, int j){
    int temp = vector[j];
    vector[j] = vector[i];
    vector[i] = temp;
}

int partition(std::vector<int> &vector, int start, int end){
    int pivot = vector[end];//pivot element
    int i = (start - 1);
    
    for(int j = start; j <= end - 1; j++){
        //if current element is smaller than the pivot
        if(vector[j] < pivot){
            i++;//increment index of smaller element
            swap(vector,i,j);
        }
    }
    i++;
    swap(vector, i,end);
    return i;
}

 /*
 Time O(n*logn) | Space O(log n)
 Not stable
 Not adaptive
 */
void quickSort(std::vector<int> &vector, int start, int end){
    if(start < end){
        int p = partition(vector, start, end); //p is partitioning index
        quickSort(vector, start, p-1);
        quickSort(vector, p+1, end);
    }
}


int main(int argc, const char * argv[]) {
    std::vector<int> testCase = {5,1,4,2,8};
    std::cout << "Input = ";
    printVector(testCase);
    
    
    quickSort(testCase,0, static_cast<int>(testCase.size() - 1));
    std::cout << "Result = ";
    printVector(testCase);
    return 0;
}

File: Sorting Algorithms/Java/QuickSort.java
import java.util.Arrays;
public class QuickSort {
    
static int partition (int a[], int start, int end)  
{  
    int pivot = a[end];//pivot element  
    int i = (start - 1);  
  
    for (int j = start; j < end ; j++)  
    {  
        //if current element is smaller than the pivot  
        if (a[j] < pivot)  
        {  
            i++; // increment index of smaller element  
            swap(a, i, j);
        }  
    }  
    i++;
    swap(a, i, end);
    return (i);  
}  
  
/*
 Time O(n*logn) | Space O(log n)
 Not stable
 Not adaptive
 */
static void quickSort(int array[], int start, int end) 
{  
    if (start < end)  
    {  
        int p = partition(array, start, end);//p is partitioning index  
        quickSort(array, start, p - 1);  
        quickSort(array, p + 1, end);  
    }  
}  

    private static void swap(int[] array, int i, int j){
        int temp = array[j];
        array[j] = array[i];
        array[i] = temp;
    }

    public static void main(String[] args) {
        int[] test = {1,2,5,-3,10};
        System.out.println("Input = " + Arrays.toString(test));
        quickSort(test, 0, test.length-1);
        System.out.println("Output = " + Arrays.toString(test));
    }
}